cmd/activator/round_trippers.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2018 The Knative Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	h2cutil "github.com/knative/serving/pkg/h2c"
+	"go.uber.org/zap"
+)
+
+type roundTripperFunc func(*http.Request) (*http.Response, error)
+
+func (rt roundTripperFunc) RoundTrip(r *http.Request) (*http.Response, error) {
+	return rt(r)
+}
+
+// baseTransport will use the appropriate transport for the request's http protocol version
+var baseTransport http.RoundTripper = roundTripperFunc(func(r *http.Request) (*http.Response, error) {
+	var transport http.RoundTripper = http.DefaultTransport
+	if r.ProtoMajor == 2 {
+		transport = h2cutil.DefaultTransport
+	}
+
+	return transport.RoundTrip(r)
+})
+
+// statusFilterRoundTripper returns an error if the response contains one of the filtered statuses.
+type statusFilterRoundTripper struct {
+	transport http.RoundTripper
+	statuses  []int
+}
+
+func newStatusFilterRoundTripper(rt http.RoundTripper, statuses ...int) http.RoundTripper {
+	return &statusFilterRoundTripper{rt, statuses}
+}
+
+func (rt *statusFilterRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	resp, err := rt.transport.RoundTrip(r)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, status := range rt.statuses {
+		if resp.StatusCode == status {
+			resp.Body.Close()
+
+			return nil, fmt.Errorf("Filtering %d", status)
```

Comment:
Converting an HTTP response to error doesn't seem right. This will definitely confuse people who are debugging this.

Instead of making this two roundtrippers that are chained together, can you simply get a func as an input: something like `type ShouldRetryFunc func(*http.Response) (bool)` and removing the `statusFilterRoundTripper` in favor of this func.

Reply:
Yeah, that makes sense. I'll take a stab at this.